File: labs/lab01/main.py
import cv2
import numpy as np
import matplotlib.pyplot as plt

def show_histogram(frame):
    """Hiển thị histogram của khung hình."""
    hsv = cv2.cvtColor(frame, cv2.COLOR_BGR2HSV)

    h, s, v = cv2.split(hsv)

    hist_h = cv2.calcHist([h], [0], None, [256], [0, 256])
    hist_s = cv2.calcHist([s], [0], None, [256], [0, 256])
    hist_v = cv2.calcHist([v], [0], None, [256], [0, 256])

    hist_h = np.array(hist_h)[:, 0]
    hist_s = np.array(hist_s)[:, 0]
    hist_v = np.array(hist_v)[:, 0]

    plt.plot(hist_h, color='b', label='Hue')
    plt.plot(hist_s, color='g', label='Saturation')
    plt.plot(hist_v, color='r', label='Value')
    plt.legend()
    plt.show()

# Chọn nguồn video là một video có sẵn trong PC
cap = cv2.VideoCapture('labs/lab01/2024-05-28 18-47-25.mp4')

while True:
  # Đọc khung hình tiếp theo
  ret, frame = cap.read()

  # Hiển thị khung hình
  cv2.imshow('Frame', frame)

  # Hiển thị histogram
  show_histogram(frame)

  # Nhấn 'q' để thoát
  if cv2.waitKey(1) & 0xFF == ord('q'):
    break

# Giải phóng nguồn
cap.release()
cv2.destroyAllWindows()

File: labs/lab01/main.cpp
#include <opencv2/opencv.hpp>

using namespace cv;

int main() {
    // Mở tệp video hoặc camera
    VideoCapture cap("2024-05-28 18-47-25.mp4"); // Thay đổi thành 0 nếu muốn sử dụng camera

    // Kiểm tra nếu video được mở thành công
    if (!cap.isOpened()) {
        std::cerr << "Không thể mở tệp video hoặc kết nối với camera!" << std::endl;
        return -1;
    }

    Mat frame;

    // Vòng lặp để đọc từng khung hình
    while (true) {
        // Đọc một khung hình từ video
        cap >> frame;

        // Kiểm tra xem có còn khung hình nào không
        if (frame.empty()) {
            std::cerr << "Không thể đọc khung hình từ video!" << std::endl;
            break;
        }

        // Tính toán histogram của khung hình
        Mat hist;
        int channels[] = {0, 1, 2}; // Tính histogram cho tất cả các kênh màu
        int histSize[] = {256, 256, 256}; // Kích thước của histogram
        float range[] = {0, 256}; // Phạm vi của giá trị pixel
        const float* ranges[] = {range, range, range};
        calcHist(&frame, 1, channels, Mat(), hist, 3, histSize, ranges, true, false);

        // Vẽ histogram
        int hist_w = 512, hist_h = 400;
        int bin_w = cvRound((double) hist_w / 256);
        Mat histImage(hist_h, hist_w, CV_8UC3, Scalar(0, 0, 0));

        // Chuẩn hóa histogram để nó vẽ trong phạm vi của hình
        normalize(hist, hist, 0, histImage.rows, NORM_MINMAX, -1, Mat());

        // Vẽ histogram cho mỗi kênh màu
        for (int i = 1; i < 256; i++) {
            line(histImage, Point(bin_w * (i - 1), hist_h - cvRound(hist.at<float>(i - 1))),
                 Point(bin_w * (i), hist_h - cvRound(hist.at<float>(i))),
                 Scalar(255, 255, 255), 2, 8, 0);
        }

        // Hiển thị khung hình và histogram
        imshow("Khung hình", frame);
        imshow("Histogram", histImage);

        // Thoát nếu nhấn phím 'q'
        if (waitKey(30) == 'q') {
            break;
        }
    }

    // Giải phóng bộ nhớ và đóng video
    cap.release();
    destroyAllWindows();

    return 0;
}
